protocol/src/main/java/org/threadly/litesockets/protocols/http/request/HTTPRequest.java

```diff
@@ -0,0 +1,71 @@
+package org.threadly.litesockets.protocols.http.request;
+
+import java.nio.ByteBuffer;
+
+import org.threadly.litesockets.protocols.http.shared.HTTPConstants;
+import org.threadly.litesockets.protocols.http.shared.HTTPHeaders;
+
+/**
+ * This is an immutable HTTPRequest object.  This is what is sent to the server when doing an
+ * HTTPRequest.  This object is created via {@link HTTPRequestBuilder}. 
+ */
+public class HTTPRequest {
+  private final HTTPRequestHeader request;
+  private final HTTPHeaders headers;
+  
+  protected HTTPRequest(HTTPRequestHeader request, HTTPHeaders headers) {
+    this.request = request;
+    this.headers = headers;
+  }
+  
+  public HTTPHeaders getHTTPHeaders() {
+    return headers;
+  }
+  
+  public HTTPRequestHeader getHTTPRequestHeaders() {
+    return request;
+  }
+  
+  public ByteBuffer getByteBuffer() {
+    ByteBuffer combined = ByteBuffer.allocate(headers.toString().length() + request.length() + 
+        HTTPConstants.HTTP_NEWLINE_DELIMINATOR.length() + 
+        HTTPConstants.HTTP_NEWLINE_DELIMINATOR.length());
+    
+    combined.put(request.getByteBuffer());
+    combined.put(HTTPConstants.HTTP_NEWLINE_DELIMINATOR.getBytes());
+    combined.put(headers.toString().getBytes());
+    combined.put(HTTPConstants.HTTP_NEWLINE_DELIMINATOR.getBytes());
+    combined.flip();
+    return combined;
+  }
+  
+  @Override
+  public String toString() {
+    return request.toString()+
+        HTTPConstants.HTTP_NEWLINE_DELIMINATOR+
+        headers.toString()+
+        HTTPConstants.HTTP_NEWLINE_DELIMINATOR;
+  }
+  
+  @Override
+  public int hashCode() {
+    return request.hashCode() ^ headers.hashCode();
+  }
+  
+  @Override
+  public boolean equals(Object o) {
```

Comment:
actually that does not work at all... maybe I am doing this wrong or something but since these might match the internal data but be a different instance this wont work.  The second I add a == check my first assert fails on the unit test for this:

    HTTPResponse hr1 = hrb.build();
    HTTPResponse hr2 = hrb.build();
    assertEquals(hr1, hr2);